firmware/library/L1_Drivers/can.hpp

```diff
@@ -1,49 +1,828 @@
+  }
+
+  void EnableBus() override
+  {
+    SetControllerMode(kReset, false);
+  }
+
+  [[gnu::always_inline]] static void ProcessIrq()
+  {
+    TxMessage_t * message_ptr;
+    bool transmit_interrupt, receive_interrupt;
+
+    for (uint8_t controller = 0; controller < kNumberOfControllers;
+         controller++)
+    {
+      if (is_controller_initialized[controller] == true)
+      {
+        // Capture interrupt register
+        // Bits are cleared when you read the ICR register
+        interrupts[controller].ICR = can_registers[controller]->ICR;
+
+        transmit_interrupt = interrupts[controller].flags.tx_buffer_1_ready ||
+                             interrupts[controller].flags.tx_buffer_2_ready ||
+                             interrupts[controller].flags.tx_buffer_3_ready;
+
+        receive_interrupt = interrupts[controller].flags.rx_buffer_full;
+
+        if (transmit_interrupt == true)
+        {
+          if (xQueueIsQueueEmptyFromISR(transmit_queue[controller]) == pdFALSE)
+          {
+            // dequeue message pointer
+            xQueueReceiveFromISR(transmit_queue[controller], &message_ptr, 0);
+            // send it to the available buffer
+            if (interrupts[controller].flags.tx_buffer_1_ready)
+            {
+              WriteMessageToBuffer(message_ptr, kBuffer1,
+                                   static_cast<Controllers>(controller));
+            }
+            else if (interrupts[controller].flags.tx_buffer_2_ready)
+            {
+              WriteMessageToBuffer(message_ptr, kBuffer2,
+                                   static_cast<Controllers>(controller));
+            }
+            else if (interrupts[controller].flags.tx_buffer_3_ready)
+            {
+              WriteMessageToBuffer(message_ptr, kBuffer3,
+                                   static_cast<Controllers>(controller));
+            }
+            else
+            {
+              // will never occur
+            }
+          }
+        }
+
+        if (receive_interrupt == true)
+        {
+          ReadMessageFromBuffer(static_cast<Controllers>(controller));
+        }
+      }
+    }
+  }
+
+  static void CanIrqHandler()
+  {
+    ProcessIrq();
+  }
+
+  template <typename T>
+  void CreateStaticQueues(T static_queue_config)
+  {
+    transmit_queue[controller_] =
+        xQueueCreateStatic(static_queue_config.kTxQueueLength,
+                           static_queue_config.kTxQueueItemSize,
+                           static_queue_config.tx_queue_storage_area,
+                           &static_queue_config.tx_static_queue);
+    receive_queue[controller_] =
+        xQueueCreateStatic(static_queue_config.kRxQueueLength,
+                           static_queue_config.kRxQueueItemSize,
+                           static_queue_config.rx_queue_storage_area,
+                           &static_queue_config.rx_static_queue);
+  }
+
+ private:
+  Controllers controller_;
```

Comment:
member variables should be at the very bottom of the file.

Reply:
done